src/psyclone/domain/lfric/kernel_interface.py

```diff
@@ -0,0 +1,841 @@
+                arg = mapping["gh_quadrature_face"](
+                    basis_tag, [int(first_dim_value_func(function_space)),
+                                ndf_symbol, nqp, nfaces],
+                    fs_name, interface=self._read_access)
+            elif shape == "gh_quadrature_edge":
+                nedges = self._create_symbol("nedges", NumberOfEdgesDataSymbol)
+                nqp = self._create_symbol("nqp", NumberOfQrPointsDataSymbol)
+                arg = mapping["gh_quadrature_edge"](
+                    basis_tag, [int(first_dim_value_func(function_space)),
+                                ndf_symbol, nqp, nedges],
+                    fs_name, interface=self._read_access)
+            elif shape in VALID_EVALUATOR_SHAPES:
+                # Need a diff basis array for each target space upon
+                # which the basis functions have been
+                # evaluated. _kern.eval_targets is a dict where the
+                # values are 2-tuples of (FunctionSpace, argument).
+                for _, _ in self._kern.eval_targets.items():
+                    raise NotImplementedError(
+                        "Evaluator shapes not implemented in kernel_interface "
+                        "class.")
+            else:
+                raise InternalError(
+                    "Unrecognised quadrature or evaluator shape '{0}'. "
+                    "Expected one of: {1}.".format(
+                        shape, VALID_EVALUATOR_SHAPES))
+            self._symbol_table.add(arg)
+            self._arglist.append(arg)
+
+    @staticmethod
+    def _basis_first_dim_value(function_space):
+        '''Internal utility to return the size of the first dimension of a
+        basis function.
+
+        :param function_space: the function space the basis function is on.
+        :type function_space: :py:class:`psyclone.domain.lfric.FunctionSpace`
+
+        :returns: an integer size as a string.
+        :rtype: str
+
+        :raises GenerationError: if an unsupported function space is \
+            supplied (e.g. ANY_SPACE_*, ANY_DISCONTINUOUS_SPACE_*)
+
+        '''
+        if function_space.has_scalar_basis:
+            first_dim = "1"
+        elif function_space.has_vector_basis:
+            first_dim = "3"
+        else:
+            # It is not possible to determine explicitly the first basis
+            # function array dimension from the metadata for any_space or
+            # any_discontinuous_space. This information needs to be passed
+            # from the PSy layer to the kernels (see issue #461).
+            raise GenerationError(
+                "Unsupported space for basis function, "
+                "expecting one of {0} but found "
+                "'{1}'".format(FunctionSpace.VALID_FUNCTION_SPACES,
+                               function_space.orig_name))
+        return first_dim
+
+    @staticmethod
+    def _diff_basis_first_dim_value(function_space):
```

Comment:
Unless I'm missing something, the implementation of this routine is identical to that for `_basis_first_dim_value`?

Reply:
It's different. In one case we call `function_space.has_scalar_basis` and `function_space.has_vector_basis` and in the other we call `function_space.has_scalar_diff_basis` and `function_space.has_vector_diff_basis`.

I agree that this is not very well written, I just copied this code across from dynamo0p3.py - which, now you know, is probably something you will want to comment on :-)

Reply:
Since it's a static method, could you just call `DynamoBasisFuntions._diff_basis_first_dim_value()` instead of duplicating the implementation? I guess we'd have to drop the leading underscore as it would then no longer be a 'private' method.

Reply:
Good point. Eventually they will probably move to this file from dynamo0p3.py but for the moment I've done as you suggest. As expected this produces circular dependencies so I've included locally - which results in the usual pylint complaint.